js/file_provider/file_provider.array.js

```diff
@@ -0,0 +1,141 @@
+import { each } from "../core/utils/iterator";
+
+import { FileProvider, FileManagerItem } from "./file_provider";
+
+class ArrayFileProvider extends FileProvider {
+
+    constructor(data) {
+        super();
+        this._data = data || [];
+    }
+
+    getItems(path, itemType) {
+        return this._getItems(path, itemType);
+    }
+
+    renameItem(item, name) {
+        item.dataItem.name = name;
+    }
+
+    createFolder(parentFolder, name) {
+        const newItem = {
+            name: name,
+            isFolder: true
+        };
+        const array = this._getChildrenArray(parentFolder.dataItem);
+        array.push(newItem);
+    }
+
+    deleteItems(items) {
+        each(items, (_, item) => this._deleteItem(item));
+    }
+
+    moveItems(items, destinationFolder) {
+        const array = this._getChildrenArray(destinationFolder.dataItem);
+        each(items, (_, item) => {
+            this._deleteItem(item);
+            array.push(item.dataItem);
+        });
+    }
+
+    copyItems(items, destinationFolder) {
+        const array = this._getChildrenArray(destinationFolder.dataItem);
+        each(items, (_, item) => {
+            const copiedItem = this._createCopy(item.dataItem);
+            array.push(copiedItem);
+        });
+    }
+
+    _createCopy(dataItem) {
+        const result = {
+            name: dataItem.name,
+            isFolder: dataItem.isFolder
+        };
+        if(dataItem.children) {
+            result.children = [];
+            each(dataItem.children, (_, childItem) => {
+                const childCopy = this._createCopy(childItem);
+                result.children.push(childCopy);
+            });
+        }
+        return result;
+    }
+
+    _deleteItem(item) {
+        let array = this._data;
+        if(item.parentPath !== "") {
+            const entry = this._findItem(item.parentPath);
+            array = entry.children;
+        }
+        const index = array.indexOf(item.dataItem);
+        array.splice(index, 1);
+    }
+
+    _getChildrenArray(dataItem) {
+        let array = null;
+        if(!dataItem) {
+            array = this._data;
+        } else {
+            if(!dataItem.children) {
+                dataItem.children = [];
+            }
+            array = dataItem.children;
```

Comment:
```suggestion
            array = ensureDefined(dataItem.children, []);
```
[core/utils/common/ensureDefined](https://github.com/DevExpress/DevExtreme/blob/19_1/js/core/utils/common.js#L9)